build/Shared/SharedExtensions.cs

```diff
@@ -126,5 +125,49 @@ public static void AppendInt(this StringBuilder sb, int value)
                 divisor /= 10;
             }
         }
+
+        public static int SingleIndex<T>(this IEnumerable<T> lst, Predicate<T> isMatch)
+        {
+            var index = 0;
+            var foundIndex = -1;
+
+            foreach (var item in lst)
+            {
+                if (isMatch(item))
+                {
+                    if (foundIndex != -1)
+                    {
+                        throw new InvalidOperationException();
+                    }
+
+                    foundIndex = index;
+                }
+
+                index++;
+            }
+
+            if (foundIndex == -1)
+            {
+                throw new InvalidOperationException();
+            }
+
+            return foundIndex;
+        }
+
+        public static int FirstIndex<T>(this IEnumerable<T> lst, Predicate<T> isMatch)
```

Comment:
Similar to https://github.com/NuGet/NuGet.Client/pull/5930/files#r1697231896


Given that this method is so far invoked on `TargetFrameworkInformation.Dependencies` which [implements IList interface](https://learn.microsoft.com/dotnet/api/system.collections.immutable.immutablearray-1?view=net-8.0), I think we could reduce the allocations here by allowing this method to be invoked on the IList and then invoke lst.NoAllocEnumerate in the foreach statement.

Reply:
Thoughts:

1) If I changed this to take in an IList, I wouldn't need the enumerator, but it would reduce the number of potential callers.

2) If I keep this as taking in an IEnumerable, it seems like there is a NoAllocEnumerate extension method defined on that too. Is there a downside to using it?

Assuming those two stmts are true, would you prefer route 1 or 2?

Reply:
Went ahead and pushed a commit where I changed the input to an IList, and just did a for loop instead of foreach